File: web-app/src/robot.js
import {FACINGS} from "./constants.js";

const clockwiseFlow = {
    [FACINGS.north]: FACINGS.east,
    [FACINGS.east]: FACINGS.south,
    [FACINGS.south]: FACINGS.west,
    [FACINGS.west]: FACINGS.north,
}

const anticlockwiseFlow = {
    [FACINGS.north]: FACINGS.west,
    [FACINGS.east]: FACINGS.north,
    [FACINGS.south]: FACINGS.east,
    [FACINGS.west]: FACINGS.south,
}

class _Robot {
    constructor(currentPosition, maxAreaToMove, facing) {
        this.currentPosition = currentPosition
        this.maxAreaToMove = maxAreaToMove
        this.facing = facing
    }

    left() {
        this.facing = anticlockwiseFlow[this.facing];
    }

    right() {
        this.facing = clockwiseFlow[this.facing];
    }

    move() {
        const mapControls = {
            [FACINGS.north]: () => this.moveToNorth(),
            [FACINGS.east]: () => this.moveToEast(),
            [FACINGS.south]: () => this.moveToSouth(),
            [FACINGS.west]: () => this.moveToWest(),
        };
        mapControls[this.facing]();
    }

    moveToNorth() {
        const currentYPosition = this.currentPosition[1];
        const newYPosition = currentYPosition - 1;

        if (newYPosition >= 0) {
            this.currentPosition[1] = newYPosition;
        }
    }

    moveToEast() {
        const currentXPosition = this.currentPosition[0];
        const maxXMoviment = this.maxAreaToMove[0] - 1;
        const newXPosition = currentXPosition + 1;

        if (newXPosition <= maxXMoviment) {
            this.currentPosition[0] = newXPosition;
        }
    }

    moveToSouth() {
        const currentYPosition = this.currentPosition[1];
        const maxYMoviment = this.maxAreaToMove[1] - 1;
        const newYPosition = currentYPosition + 1;

        if (newYPosition <= maxYMoviment) {
            this.currentPosition[1] = newYPosition;
        }
    }

    moveToWest() {
        const currentXPosition = this.currentPosition[0];
        const newXPosition = currentXPosition - 1;

        if (newXPosition >= 0) {
            this.currentPosition[0] = newXPosition;
        }
    }

    getCurrentPosition() {
        return {
            x: this.currentPosition[0],
            y: this.currentPosition[1],
            facing: this.facing,
        }
    }

    currentPositionReport() {
        console.log(`${this.currentPosition[0]},${this.currentPosition[1]},${this.facing}`)
    }
}

const Robot = (currentPosition, maxAreaToMove, facing) => {
    const robot = new _Robot(currentPosition, maxAreaToMove, facing);

    return {
        left: () => robot.left(),
        right: () => robot.right(),
        move: () => robot.move(),
        currentPositionReport: () => robot.currentPositionReport(),
        getCurrentPosition: () => robot.getCurrentPosition(),
    }
}

export default Robot;

File: cli-ruby/lib/robot.rb
CLOCKWISE_FLOW = {
  north: :east,
  east: :south,
  south: :west,
  west: :north
}

ANTICLOCKWISE_FLOW = {
  north: :west,
  east: :north,
  south: :east,
  west: :south
}

class Robot
  attr_reader :current_position
  attr_reader :max_area_to_move
  attr_reader :facing

  def initialize(attributes)
    @current_position = attributes[:initial_position]
    @facing = attributes[:facing]
    @max_area_to_move = attributes[:max_area_to_move]
  end

  def left
    @facing = ANTICLOCKWISE_FLOW[@facing]
  end

  def right
    @facing = CLOCKWISE_FLOW[@facing]
  end

  def move
    move_to_north if (@facing == :north)
    move_to_east if (@facing == :east)
    move_to_south if (@facing == :south)
    move_to_west if (@facing == :west)
  end

  def move_to_north
    current_y_position = @current_position[1]
    new_y_position = current_y_position - 1

    if new_y_position >= 0
      @current_position[1] = new_y_position
    end
  end

  def move_to_east
    current_x_position = @current_position[0]
    max_x_moviment = @max_area_to_move[0] - 1
    new_x_position = current_x_position + 1

    if new_x_position <= max_x_moviment
      @current_position[0] = new_x_position
    end
  end

  def move_to_south
    current_y_position = @current_position[1]
    max_y_moviment = @max_area_to_move[1] - 1
    new_y_position = current_y_position + 1

    if new_y_position <= max_y_moviment
      @current_position[1] = new_y_position
    end
  end

  def move_to_west
    current_x_position = @current_position[0]
    new_x_position = current_x_position - 1

    if new_x_position >= 0
      @current_position[0] = new_x_position
    end
  end

  def current_position_report
    print "#{@current_position[0]},#{@current_position[1]},#{@facing.upcase}"
  end

  private :move_to_north, :move_to_east, :move_to_south, :move_to_west
end